hibernate-rx-core/src/main/java/org/hibernate/rx/event/impl/DefaultRxLoadEventListener.java

```diff
@@ -74,27 +76,31 @@ public void onLoad(
 			throw new HibernateException( "Unable to locate persister: " + event.getEntityClassName() );
 		}
 
-		Object checkId = checkId( event, loadType, persister );
-		if ( checkId instanceof CompletionStage) {
+		CompletionStage<Void> checkId = checkId( event, loadType, persister );
+		if ( !checkId.toCompletableFuture().isDone() ) {
 			// This only happens if the object is loaded from the db
-			throw new UnsupportedOperationException( "Unexpected access to the db" );
+			throw new AssertionFailure( "Unexpected access to the database" );
 		}
 
 		try {
-			Object loaded = doOnLoad( persister, event, loadType );
-			if ( loaded instanceof CompletionStage ) {
+			CompletionStage<Optional<Object>> loaded = doOnLoad( persister, event, loadType );
+			if ( !loaded.toCompletableFuture().isDone() ) {
 				// This only happens if the object is loaded from the db
-				throw new UnsupportedOperationException( "Unexpected access to the db" );
+				throw new AssertionFailure( "Unexpected access to the database" );
 			}
 			else {
 				// Proxy
-				event.setResult( loaded );
+				loaded.toCompletableFuture().getNow( Optional.empty() ).ifPresent( event::setResult );
 			}
 		}
 		catch (HibernateException e) {
 			LOG.unableToLoadCommand( e );
 			throw e;
 		}
+
+		if ( event.getResult() instanceof CompletionStage ) {
+			throw new AssertionFailure( "Unexpected CompletionStage" );
+		}
```

Comment:
Can this happen or is it a safety check?

Reply:
It's just an assertion, to help detect bugs. It should never ever occur. (That's why I had it throw `AssertionFailure`.)